file/builder.go

```diff
@@ -890,6 +962,24 @@ func (b *stateBuilder) plugins() {
 	}
 }
 
+func (b *stateBuilder) validatePlugin(p FPlugin) error {
+	if (b.isConsumerGroupScopedPluginSupported && !b.isKonnect) && *p.Name == ratelimitingAdvancedPluginName {
+		// check if deprecated consumer-groups configuration is present in the config
+		var consumerGroupsFound bool
+		if groups, ok := p.Config["consumer_groups"]; ok {
+			// if groups is an array of length > 0, then consumer_groups is set
+			if groupsArray, ok := groups.([]interface{}); ok && len(groupsArray) > 0 {
+				consumerGroupsFound = true
+			}
+		}
+		_, enforceConsumerGroupsFound := p.Config["enforce_consumer_groups"]
+		if consumerGroupsFound || enforceConsumerGroupsFound {
+			return utils.ErrorConsumerGroupUpgrade
+		}
+	}
+	return nil
+}
```

Comment:
I hit this validation with the config
```json
{
  "_format_version": "3.0",
  "_info": { "select_tags": ["managed-by-ingress-controller"], "defaults": {} },
  "consumers":
    [
      {
        "username": "test-consumer-2",
        "tags":
          [
            "k8s-name:test-consumer-2",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongConsumer",
            "k8s-uid:139a69b7-e1ae-49e9-b904-e2c42a248194",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1",
          ],
        "keyauth_credentials":
          [
            {
              "consumer":
                {
                  "id": "622e9ec5-64fe-492a-8bc2-d712f3aad1b1",
                  "username": "test-consumer-2",
                },
              "key": "test-consumer-2",
              "tags":
                [
                  "k8s-name:test-consumer-2",
                  "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
                  "k8s-kind:Secret",
                  "k8s-uid:7f25dc61-11eb-4fdc-a783-c825a554cd13",
                  "k8s-version:v1",
                ],
            },
          ],
        "groups":
          [
            { "name": "test-consumer-group-1" },
            { "name": "test-consumer-group-2" },
          ],
      },
    ],
  "consumer_groups":
    [
      {
        "name": "test-consumer-group-2",
        "tags":
          [
            "k8s-name:test-consumer-group-2",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongConsumerGroup",
            "k8s-uid:fea5aeed-81a0-4616-8b77-79b173023a6a",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1beta1",
          ],
      },
      {
        "name": "test-consumer-group-1",
        "tags":
          [
            "k8s-name:test-consumer-group-1",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongConsumerGroup",
            "k8s-uid:5b979d79-3a46-4aba-b10b-070ae88218bd",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1beta1",
          ],
      },
    ],
  "plugins":
    [
      {
        "name": "response-transformer-advanced",
        "config":
          {
            "add":
              {
                "headers": ["response-transformer-advanced-1: added"],
                "if_status": [],
                "json": [],
                "json_types": [],
              },
            "allow": { "json": null },
            "append":
              { "headers": [], "if_status": [], "json": [], "json_types": [] },
            "dots_in_keys": true,
            "remove": { "headers": [], "if_status": [], "json": [] },
            "rename": { "headers": [], "if_status": [] },
            "replace":
              {
                "body": null,
                "headers": [],
                "if_status": [],
                "json": [],
                "json_types": [],
              },
            "transform": { "functions": [], "if_status": [], "json": [] },
          },
        "consumer_group": "test-consumer-group-2",
        "enabled": true,
        "protocols": ["grpc", "grpcs", "http", "https"],
        "tags":
          [
            "k8s-name:response-transformer-advanced-1",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongPlugin",
            "k8s-uid:328e096a-3f4b-46c3-acd4-18cf871193a5",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1",
          ],
      },
      {
        "name": "response-transformer-advanced",
        "config":
          {
            "add":
              {
                "headers": ["response-transformer-advanced-1: added"],
                "if_status": [],
                "json": [],
                "json_types": [],
              },
            "allow": { "json": null },
            "append":
              { "headers": [], "if_status": [], "json": [], "json_types": [] },
            "dots_in_keys": true,
            "remove": { "headers": [], "if_status": [], "json": [] },
            "rename": { "headers": [], "if_status": [] },
            "replace":
              {
                "body": null,
                "headers": [],
                "if_status": [],
                "json": [],
                "json_types": [],
              },
            "transform": { "functions": [], "if_status": [], "json": [] },
          },
        "consumer_group": "test-consumer-group-1",
        "enabled": true,
        "protocols": ["grpc", "grpcs", "http", "https"],
        "tags":
          [
            "k8s-name:response-transformer-advanced-1",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongPlugin",
            "k8s-uid:328e096a-3f4b-46c3-acd4-18cf871193a5",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1",
          ],
      },
      {
        "name": "rate-limiting-advanced",
        "config":
          {
            "consumer_groups": null,
            "dictionary_name": "kong_rate_limiting_counters",
            "disable_penalty": false,
            "enforce_consumer_groups": false,
            "error_code": 429,
            "error_message": "API rate limit exceeded",
            "header_name": null,
            "hide_client_headers": false,
            "identifier": "consumer",
            "limit": [100],
            "namespace": null,
            "path": null,
            "redis":
              {
                "cluster_addresses": null,
                "connect_timeout": null,
                "database": 0,
                "host": null,
                "keepalive_backlog": null,
                "keepalive_pool_size": 256,
                "password": null,
                "port": null,
                "read_timeout": null,
                "send_timeout": null,
                "sentinel_addresses": null,
                "sentinel_master": null,
                "sentinel_password": null,
                "sentinel_role": null,
                "sentinel_username": null,
                "server_name": null,
                "ssl": false,
                "ssl_verify": false,
                "timeout": 2000,
                "username": null,
              },
            "retry_after_jitter_max": 0,
            "strategy": "local",
            "sync_rate": null,
            "window_size": [100],
            "window_type": "sliding",
          },
        "consumer_group": "test-consumer-group-2",
        "enabled": true,
        "protocols": ["grpc", "grpcs", "http", "https"],
        "tags":
          [
            "k8s-name:rate-limiting-advanced-1",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongPlugin",
            "k8s-uid:3f689639-3a59-4e94-ba2d-b698a7cbc63f",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1",
          ],
      },
      {
        "name": "key-auth",
        "config":
          {
            "anonymous": null,
            "hide_credentials": false,
            "key_in_body": false,
            "key_in_header": true,
            "key_in_query": true,
            "key_names": ["apikey"],
            "run_on_preflight": true,
          },
        "route": "433a5562-833d-41aa-a458-cf9e827b4cdf.echo.echo..1027",
        "enabled": true,
        "protocols": ["grpc", "grpcs", "http", "https", "ws", "wss"],
        "tags":
          [
            "k8s-name:key-auth",
            "k8s-namespace:433a5562-833d-41aa-a458-cf9e827b4cdf",
            "k8s-kind:KongPlugin",
            "k8s-uid:5d2a6858-5f41-410e-ae19-83ab2f6e796a",
            "k8s-group:configuration.konghq.com",
            "k8s-version:v1",
          ],
      },
    ],
}
```
for `response-transformer-advanced` everything works like a charm.

The problem is for `rate-limiting-advanced` what does it mean
> Please use Consumer Groups scoped Plugins when running against Kong Enterprise 3.4.0 and above.

Is there a different `rate-limiting-advanced`?

Reply:
Would you mind giving it another try now?

> Is there a different rate-limiting-advanced?

Not sure what you mean here 🤔